### DISCUSSION TITLE
Transistions for specific classes

### DISCUSSION THREAD

**Discussion by sards3** [AvaloniaUI/Avalonia#16345] [Q&A] [answered]
On my control, I want to transition the `Background` property when a specific class is applied, but not when the `:pointerover` pseudoclass is applied. Is there a way to do this? Alternately, is there a way to write a `Transition<T>` with custom logic for when to apply it? It seems like I could do it if I could override `Transition<T>.Apply(...)`, but unfortunately that method is `internal`.  It would be nice if we could make that method `protected`.

Here's an example:

```XML
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:Class="Gui.Tempaxaml">
	<UserControl.Styles>
		<Style Selector="Border#TestBorder">
			<Setter Property="Transitions">
				<Transitions>
					<BrushTransition Property="Background" Duration="0:0:1" />
				</Transitions>
			</Setter>
			<Setter Property="Background" Value="Red" />
			<Style Selector="^.Yellow">
				<!-- I want a transition here. -->
				<Setter Property="Background" Value="Yellow"/>
			</Style>
			<Style Selector="^:pointerover">
				<!-- But I don't want a transition here. -->
				<Setter Property="Background" Value="Blue"/>				
			</Style>		    	
		</Style>
	</UserControl.Styles>
	<StackPanel>
		<CheckBox Name="TestCheckbox">Apply 'Yellow' class</CheckBox>
		<Border Name="TestBorder" Width="200" Height="200" Classes.Yellow="{Binding #TestCheckbox.IsChecked}" />	
	</StackPanel>	
</UserControl>
```

**Comment by stevemonaco** [comment, marked as answer]
You can try separating the selectors, but it's still problematic. eg.

```xml
<UserControl.Styles>
    <Style Selector="Border#TestBorder">
        <Setter Property="Transitions">
            <Transitions>
                <BrushTransition Property="Background" Duration="0:0:1" />
            </Transitions>
        </Setter>
        <Setter Property="Background" Value="Red" />
        <Style Selector="^.Yellow">
            <!--  I want a transition here.  -->
            <Setter Property="Background" Value="Yellow" />
        </Style>
    </Style>
    <Style Selector="Border#TestBorder:pointerover">
        <!--  But I don't want a transition here.  -->
        <Setter Property="Transitions" Value="{x:Null}" />
        <Setter Property="Background" Value="Blue" />
    </Style>
</UserControl.Styles>
```

This sets `Background=Blue` immediately when `:pointerover`. The issue is that when the mouse exits, the `Background` will transition slowly back to Red or immediately to Yellow. The state transitions (starting and ending states) aren't defined clearly enough here and Avalonia currently doesn't support enter/exit states, AFAIK.

**Reply by sards3** [reply]
Thanks; this is good enough for my purposes. I still would like to be able to override `Transition<T>.Apply()` to write custom logic.

**Reply by stevemonaco** [reply]
AFAIK, you can write your own transitions (like BrushTransition) but the process may be complex. I think the docs are sparse here and there's not a lot of guidance of `Transition`s vs `Animation`s. I haven't had the time/need to study deeper.
